### DISCUSSION TITLE
Which JDK versions should we support?

### DISCUSSION THREAD

**Discussion by ftomassetti** [Strumenta/kolasu#103] [General]
Version 1.5 is aiming to JDK 11. Is this ok or should we support JDK 8 instead?

**Comment by alessiostalla** [comment]
Just for information, ANTLR 4.10+ requires Java 11.

**Reply by ftomassetti** [reply]
That seems a good point in favor of Java 11

**Reply by alessiostalla** [reply]
Maybe. However if we want to decouple Kolasu from ANTLR that becomes less important.
In general we have some clients that require Java 8 so that could mean that if we move Kolasu 1.5 to Java 11 we would have to maintain Kolasu 1.4 for a longer period and we would have some parsers that are stuck on Kolasu 1.4 (unless we also want to maintain 2 versions of those parsers)

**Comment by ftomassetti** [comment]
In 2020 Java 8 was still used a lot: https://snyk.io/blog/developers-dont-want-to-leave-java-8-as-64-hold-firm-on-their-preferred-release/

**Comment by ftomassetti** [comment]
In 2022 Java 8 is still used by over 40% of companies in production. I think this means we should support it. https://newrelic.com/resources/report/2022-state-of-java-ecosystem

**Comment by ftomassetti** [comment]
The only reason I can think of for supporting Java 11, is for getting to use ANTLR 4.11. It may introduce some performance advantages and support for more targets. That said if Java 8 still used by over 40% of companies that seem something we cannot ignore. Supporting two variants of each of our parsers seem unsustainable.

Maybe decoupling ANTLR from kolasu-core could help us here. We could potentially have a modules such as: kolasu-antlr49 (using Java 8) and kolasu-antlr411 (using Java 11). We would use kolasu-antlr49 unless strong reasons push as towards using kolasu-antlr411. 

@alessiostalla do you think this may work?

**Reply by alessiostalla** [reply]
Yes but it's probably not necessary. We can have Kolasu depend on ANTLR 4.9, while projects using Kolasu can depend on a more recent version as they see fit, as long as ANTLR doesn't change their APIs significantly.

**Reply by ftomassetti** [reply]
Ok, but if in a project one wants to use ANTLR 4.10 and Kolasu depends on ANTLR 4.9, will Gradle sort it out and take the most recent one? Will the dependency used directly in the project prevail on the indirect dependency coming from Kolasu? I am confused by this sort of things

**Reply by alessiostalla** [reply]
I know that with Maven if you explicitly ask for 4.10 in the dependent project you'll get 4.10 regardless of what Kolasu specifies. I don't know about Gradle but I would be surprised if it didn't allow this.

**Reply by ftomassetti** [reply]
Ok, then I think we agree to move to Java 8 and ANTLR 4.9.3 in Kolasu

**Comment by ftomassetti** [comment]
Kolasu 1.5 has been moved to Java 8
